RgbdCamera: Improve rendering speed

kunimatsu-tri opened the pull request:
This PR improves rendering speed by eliminating duplicate calls for VTK update.

The former `PerformVTKUpdate` becomes `UpdateModelPoses` and
The former `UpdateRenderingWIndow` becomes `PerformVTKUpdate` now so that the function names reflect exactly what they are doing.

The computational speed of each of `OutputColorImage`, `OutputDepthImage` and `OutputLabelImage` improved from roughly 40 ms to 20 ms.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6399)
<!-- Reviewable:end -->

kunimatsu-tri commented:
+@ericcousineau-tri feature review please. Thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-KnAPSopdKIepon4_Hnk:b-6wg1k9)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnG_bfZ2SkkMNfR_lL1:b-2ze9bv) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
> ```cpp
>       for (auto& id_object_map : id_object_maps_) {
>         auto& actor = id_object_map.at(body->get_body_index()).at(i);
>         actor->SetUserTransform(vtk_transform);
> ```

BTW Is it possible to put a TODO to just transform the VTK camera rather than transform each object?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-KnGai5JGSo6ITlQAozF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnZOxh2zoLP6UjJzd6V:b-5q20e0) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Is it possible to put a TODO to just transform the VTK camera rather than transform each object?
</blockquote></details>

I believe this is the most appropriate way to render scenes with VTK.  As you can see in the [OpenGL tutorial](http://www.opengl-tutorial.org/beginners-tutorials/tutorial-3-matrices/), the way they render is moving all the objects into the camera coordinate system. 

Actually I tried moving `vtkCamera` before. However, due to the stateful behavior, I had to do weird thing (reset the camera by setting unit matrix first, then set the transformation matrix).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
+@sherm1 platform review please. Thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-KnZQ0BZBBU7uO-j4XTx:bqs0osh)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I couldn't help but stick my nose in....I added a couple of BTW.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 688 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnZdjvpIHoCei6eAYy8:-KnZdjvpIHoCei6eAYy9:b-e6owe4) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L688)):*
> ```cpp
>     for (size_t i = 0; i < body->get_visual_elements().size(); ++i) {
>       const auto& visual = body->get_visual_elements()[i];
>       const auto X_CVisual = X_CW * tree_.CalcBodyPoseInWorldFrame(
> ```

BTW You're doing redundant matrix multiplication here.  `X_CW * tree_.CalcBodyPoseInWorldFrame(cache, *body)` should be multiplied once in the *outer* for loop and only the concatenation of `X_CB` (body to camera) and `X_BE = visual.getLocalTransform()` (element to body) should be multiplied in this inner loop). You might not recognize benefit in your tests -- because it's value is only really manifest when a body has multiple visual elements -- but it is worthwhile.

---

*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnZeYdLfGk8wtJi3u-W:b-6ewtwb) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I believe this is the most appropriate way to render scenes with VTK.  As you can see in the [OpenGL tutorial](http://www.opengl-tutorial.org/beginners-tutorials/tutorial-3-matrices/), the way they render is moving all the objects into the camera coordinate system. 

Actually I tried moving `vtkCamera` before. However, due to the stateful behavior, I had to do weird thing (reset the camera by setting unit matrix first, then set the transformation matrix).
</blockquote></details>

BTW I generally agree with @ericcousineau-tri's point. The fact that you have a camera *and* are transforming things into camera space seems suspicious. Usually the existence of a camera implies an automation that should do that for you.

---

*[drake/systems/sensors/rgbd_camera.cc, line 703 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnZn-_2gM5_qXpW8FQ8:-KnZn-_2gM5_qXpW8FQ9:bjmai8z) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L703)):*
> ```cpp
>   }
> 
>   if (!kCameraFixed) {
> ```

BTW Related to @ericcousineau-tri's earlier comment. The fact that you are "re-posing" terrain is definitely a warning sign.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-KnZn9WGRyhn1yN8Ep6C:b-tuhzli)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 688 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnZdjvpIHoCei6eAYy8:-KnZrFcSIjj-8XzmVNO_:b-x8luaj) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L688)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW You're doing redundant matrix multiplication here.  `X_CW * tree_.CalcBodyPoseInWorldFrame(cache, *body)` should be multiplied once in the *outer* for loop and only the concatenation of `X_CB` (body to camera) and `X_BE = visual.getLocalTransform()` (element to body) should be multiplied in this inner loop). You might not recognize benefit in your tests -- because it's value is only really manifest when a body has multiple visual elements -- but it is worthwhile.
</blockquote></details>

Good catch!  Will change accordingly.

---

*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnZtCplHarfCLL3ODgX:b-2okwpq) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I generally agree with @ericcousineau-tri's point. The fact that you have a camera *and* are transforming things into camera space seems suspicious. Usually the existence of a camera implies an automation that should do that for you.
</blockquote></details>

To be clear, I also agree with Eric's point in general.  The reason I didn't do it is that  I found `vtkCamera`'s behavior is weird.

Please see the [doc](http://www.vtk.org/doc/release/5.8/html/a00281.html#1c787ee7f1a70fe3203aa60ea041b4b1):
```
void vtkCamera::ApplyTransform(vtkTransform *t) 	
Apply a transform to the camera. The camera position, focal-point, and view-up are re-calculated using the transform's matrix to multiply the old points by the new transform.
```

This means you have to either 1) reset camera pose first, and then set transformation matrix or 2) get current camera matrix first, then multiply the inverse of the current matrix to new transformation matrix.

paging @sankhesh: Can you suggest what the right thing to do this with VTK?

---

*[drake/systems/sensors/rgbd_camera.cc, line 703 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnZn-_2gM5_qXpW8FQ8:-KnZx58rfnXutf98qnE0:bm88x07) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L703)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Related to @ericcousineau-tri's earlier comment. The fact that you are "re-posing" terrain is definitely a warning sign.
</blockquote></details>

Let's wait for hearing from @sankhesh.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnZxpeY-dmot-sym7j9:b-c0gjrq) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

To be clear, I also agree with Eric's point in general.  The reason I didn't do it is that  I found `vtkCamera`'s behavior is weird.

Please see the [doc](http://www.vtk.org/doc/release/5.8/html/a00281.html#1c787ee7f1a70fe3203aa60ea041b4b1):
```
void vtkCamera::ApplyTransform(vtkTransform *t) 	
Apply a transform to the camera. The camera position, focal-point, and view-up are re-calculated using the transform's matrix to multiply the old points by the new transform.
```

This means you have to either 1) reset camera pose first, and then set transformation matrix or 2) get current camera matrix first, then multiply the inverse of the current matrix to new transformation matrix.

paging @sankhesh: Can you suggest what the right thing to do this with VTK?

</blockquote></details>

This isn't surprising at all.  `ApplyTransform` means transform concatenation. So, it will *always* be a matrix multiplication. Borrowing from the OpenGL paradigm, the model-view matrix is treated like a matrix stack. You start with the identity and then push and pop (via `glPushMatrix` and `glPopMatrix`) up and down a hierarchy. So, it's unsurprising to see this. What you *want* to call is the `SetModelTransformMatrix` method. THis is the equivalent of `glLoadMatrix` and should replace the old transform with the new.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-Kn_-ZAnRJRQaW1MFJpx:b-jh2v41) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

This isn't surprising at all.  `ApplyTransform` means transform concatenation. So, it will *always* be a matrix multiplication. Borrowing from the OpenGL paradigm, the model-view matrix is treated like a matrix stack. You start with the identity and then push and pop (via `glPushMatrix` and `glPopMatrix`) up and down a hierarchy. So, it's unsurprising to see this. What you *want* to call is the `SetModelTransformMatrix` method. THis is the equivalent of `glLoadMatrix` and should replace the old transform with the new.
</blockquote></details>

Oh thanks. Good to know. Let me try `SetModelTransformMatrix` then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-Kn_1_zSQ9P6_uKEW58j:b55oy2i) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Oh thanks. Good to know. Let me try `SetModelTransformMatrix` then.
</blockquote></details>

There is no `SetModelTransformMatrix` method on VTK 5.8. It seems that method introduced since VTK 5.10.

I'll put TODO comment for now saying something like
```
TODO (kunimatsu-tri) Once VTK 5.8 support dropped, rewrite camera pose transformation using `SetModelTransformMatrix`.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 688 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnZdjvpIHoCei6eAYy8:-Kn_55tg2S2RQyVKweBk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L688)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Good catch!  Will change accordingly.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-Kn_57vpbemcBoFoSCa7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

There is no `SetModelTransformMatrix` method on VTK 5.8. It seems that method introduced since VTK 5.10.

I'll put TODO comment for now saying something like
```
TODO (kunimatsu-tri) Once VTK 5.8 support dropped, rewrite camera pose transformation using `SetModelTransformMatrix`.
```

</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-Kn_6TbBkx73Ow0hIxWd:b68cr26) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Even in this context, it would probably *still* be better to change the code to use the "set the identity and then call `ApplyTransform`" approach. Then, the TODO is nicely localized -- those two calls would change to one. If we leave the code as is, the for loop has to change and it will be harder to remember what *exactly* needs to be done at that point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending a few very minor BTWs.

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 216 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-Kn_OnlIitsdRF3t52tG:-Kn_OnlIitsdRF3t52tH:b-i1ru6q) ([raw file](https://github.com/robotlocomotion/drake/blob/6244ffe8f998c003be1b28e55d7be58f9ddbdbd7/drake/systems/sensors/rgbd_camera.cc#L216)):*
> ```cpp
> };
> 
> void PerformVTKUpdate(
> ```

BTW a short comment here explaining what this is doing, especially if there is anything a future programmer should be worried about when using it, would be helpful.

---

*[drake/systems/sensors/rgbd_camera.cc, line 697 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-Kn_OZoAIaGxzTv6jgRH:-Kn_OZoBDY6q9YIRYc8l:b-k9s1o7) ([raw file](https://github.com/robotlocomotion/drake/blob/6244ffe8f998c003be1b28e55d7be58f9ddbdbd7/drake/systems/sensors/rgbd_camera.cc#L697)):*
> ```cpp
>           VtkUtil::ConvertToVtkTransform(X_CVisual);
>       // `id_object_maps_` is modified here. This is OK because 1) we are just
>       // copying data to the memory spaces allocated at the construction time
> ```

BTW  nit "~~the~~ construction time" -> "construction time"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-Kn_P5g_t2cuFM6yf0Br:bhnbv0q)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-Kn_bjfyQhQQqlz38nnt:b9s8wj2) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Even in this context, it would probably *still* be better to change the code to use the "set the identity and then call `ApplyTransform`" approach. Then, the TODO is nicely localized -- those two calls would change to one. If we leave the code as is, the for loop has to change and it will be harder to remember what *exactly* needs to be done at that point.
</blockquote></details>

OK, will do in separated PR since this is unrelated to speed improvement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comments except @SeanCurtis-TRI's which will be addressed at upcoming PR.
Thank you all for reviewing.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 216 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-Kn_OnlIitsdRF3t52tG:-Kn_e3ByxrBkH_m7WTk5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6244ffe8f998c003be1b28e55d7be58f9ddbdbd7/drake/systems/sensors/rgbd_camera.cc#L216)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW a short comment here explaining what this is doing, especially if there is anything a future programmer should be worried about when using it, would be helpful.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 697 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-Kn_OZoAIaGxzTv6jgRH:-Kn_cm4iLZ3pyaWaU_Fd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6244ffe8f998c003be1b28e55d7be58f9ddbdbd7/drake/systems/sensors/rgbd_camera.cc#L697)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW  nit "~~the~~ construction time" -> "construction time"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-Kn_erB9t2V5IE1hyHXy:bl6mq1l)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
@drake-jenkins-bot mac-clang-experimental please
@drake-jenkins-bot mac-clang-ninja-experimental please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-Kn_gnQQ8z2TXHcfOzJN:bg0xcbm)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> You can have one from me as well, just so I don't inadvertently interfere.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-Kn_lvhBvMKTjFqPp7mF:b8kr8o4)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
thanks Sean :)

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-:-Kn_p5MSFkHeJOtHCsWI:bud0g5a)*
<!-- Sent from Reviewable.io -->

sankhesh commented:
*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnaAAFjoCroudHFEMB3:b-db22lm) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

OK, will do in separated PR since this is unrelated to speed improvement.
</blockquote></details>

@kunimatsu-tri Based on what I understand, you want to set model-view transform on the camera.  `vtkCamera` doesn't allow you to explicitly set the model view transform matrix, but it is computed internally using the position, focal point and view-up vector set on the camera. So, you could just set those three ivars based on the transform you choose. Alternatively, you could use `vtkExternalOpenGLCamera` which allows setting the view transform explicitly; although, you'll be adding a dependency of `vtkRenderingExternal` module to your stack. Hope that helps.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 698 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6399#-KnG_bfYwUR0RtxmzhGy:-KnachHFiMHPUEGdBC2J:baqd2w4) ([raw file](https://github.com/robotlocomotion/drake/blob/8329df451407ad28e3b29c196cc43dc1df19ca50/drake/systems/sensors/rgbd_camera.cc#L698)):*
<details><summary><i>Previously, sankhesh (Sankhesh Jhaveri) wrote…</i></summary><blockquote>

@kunimatsu-tri Based on what I understand, you want to set model-view transform on the camera.  `vtkCamera` doesn't allow you to explicitly set the model view transform matrix, but it is computed internally using the position, focal point and view-up vector set on the camera. So, you could just set those three ivars based on the transform you choose. Alternatively, you could use `vtkExternalOpenGLCamera` which allows setting the view transform explicitly; although, you'll be adding a dependency of `vtkRenderingExternal` module to your stack. Hope that helps.
</blockquote></details>

Thanks @sankhesh!

I used those three ivars together with ApplyTransform. PTAL at #6432.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6399)*
<!-- Sent from Reviewable.io -->
